File: src/ai_usage_bar/poller.py
from __future__ import annotations

import logging
import threading
import time
from collections.abc import Callable

from ai_usage_bar.config import AppConfig
from ai_usage_bar.models import ProviderSnapshot
from ai_usage_bar.providers.codex import fetch_codex
from ai_usage_bar.providers.cursor import fetch_cursor

log = logging.getLogger(__name__)


class Poller:
    def __init__(
        self,
        get_config: Callable[[], AppConfig],
        on_update: Callable[[list[ProviderSnapshot]], None],
    ) -> None:
        self._get_config = get_config
        self._on_update = on_update
        self._stop = threading.Event()
        self._wake = threading.Event()
        self._thread: threading.Thread | None = None
        self._next_ok: dict[str, float] = {"cursor": 0.0, "codex": 0.0}
        self._last: dict[str, ProviderSnapshot] = {}

    def start(self) -> None:
        if self._thread and self._thread.is_alive():
            return
        self._stop.clear()
        self._thread = threading.Thread(target=self._loop, name="ai-usage-bar-poller", daemon=True)
        self._thread.start()

    def stop(self) -> None:
        self._stop.set()
        self._wake.set()

    def refresh_now(self) -> None:
        self._next_ok["cursor"] = 0.0
        self._next_ok["codex"] = 0.0
        self._wake.set()

    def _loop(self) -> None:
        while not self._stop.is_set():
            try:
                snapshots = self._collect()
                self._on_update(snapshots)
            except Exception:
                log.exception("使用量の更新処理でエラー")
            config = self._get_config()
            self._wake.wait(timeout=max(30, config.refresh_seconds))
            self._wake.clear()

    def _collect(self) -> list[ProviderSnapshot]:
        config = self._get_config()
        now = time.monotonic()
        snapshots: list[ProviderSnapshot] = []
        if config.show_cursor:
            snapshots.append(self._fetch_one("cursor", now, fetch_cursor))
        if config.show_codex:
            snapshots.append(self._fetch_one("codex", now, fetch_codex))
        return snapshots

    def _fetch_one(
        self,
        key: str,
        now: float,
        fetcher: Callable[[], ProviderSnapshot],
    ) -> ProviderSnapshot:
        cached = self._last.get(key)
        if now < self._next_ok[key] and cached is not None:
            return cached
        snapshot = fetcher()
        self._last[key] = snapshot
        delay = 0.0
        if snapshot.error == "レート制限":
            delay = 300.0
        elif snapshot.error == "要認証":
            delay = 180.0
        self._next_ok[key] = time.monotonic() + delay
        return snapshot

File: src/AiUsageBar.Core/Poller.cs
using AiUsageBar.Core.Providers;

namespace AiUsageBar.Core;

public sealed class Poller : IDisposable
{
    private readonly Func<AppConfig> _getConfig;
    private readonly Action<IReadOnlyList<ProviderSnapshot>> _onUpdate;
    private readonly HttpClient _http;
    private readonly CancellationTokenSource _cts = new();
    private readonly ManualResetEventSlim _wake = new(false);
    private Task? _loop;
    private readonly Dictionary<string, double> _nextOk = new() { ["cursor"] = 0, ["codex"] = 0 };
    private readonly Dictionary<string, ProviderSnapshot> _last = [];

    public Poller(Func<AppConfig> getConfig, Action<IReadOnlyList<ProviderSnapshot>> onUpdate)
    {
        _getConfig = getConfig;
        _onUpdate = onUpdate;
        _http = new HttpClient { Timeout = TimeSpan.FromSeconds(20) };
    }

    public void Start()
    {
        _loop ??= Task.Run(() => LoopAsync(_cts.Token));
    }

    public void Stop()
    {
        _cts.Cancel();
        _wake.Set();
    }

    public void RefreshNow()
    {
        _nextOk["cursor"] = 0;
        _nextOk["codex"] = 0;
        _wake.Set();
    }

    public void Dispose()
    {
        Stop();
        _http.Dispose();
        _cts.Dispose();
        _wake.Dispose();
    }

    private async Task LoopAsync(CancellationToken cancellationToken)
    {
        while (!cancellationToken.IsCancellationRequested)
        {
            try
            {
                var snapshots = await CollectAsync(cancellationToken).ConfigureAwait(false);
                _onUpdate(snapshots);
            }
            catch (Exception ex) when (ex is not OperationCanceledException)
            {
                AppLog.Error("使用量の更新処理でエラー", ex);
            }

            var config = _getConfig();
            try
            {
                _wake.Reset();
                _wake.Wait(TimeSpan.FromSeconds(Math.Max(30, config.RefreshSeconds)), cancellationToken);
            }
            catch (OperationCanceledException)
            {
                break;
            }
        }
    }

    private async Task<List<ProviderSnapshot>> CollectAsync(CancellationToken cancellationToken)
    {
        var config = _getConfig();
        var now = Environment.TickCount64 / 1000.0;
        var snapshots = new List<ProviderSnapshot>();
        if (config.ShowCursor)
        {
            snapshots.Add(await FetchOneAsync("cursor", now, () => CursorProvider.FetchAsync(_http, cancellationToken)).ConfigureAwait(false));
        }

        if (config.ShowCodex)
        {
            snapshots.Add(await FetchOneAsync("codex", now, () => CodexProvider.FetchAsync(_http, cancellationToken)).ConfigureAwait(false));
        }

        return snapshots;
    }

    private async Task<ProviderSnapshot> FetchOneAsync(string key, double now, Func<Task<ProviderSnapshot>> fetcher)
    {
        if (now < _nextOk[key] && _last.TryGetValue(key, out var cached))
        {
            return cached;
        }

        var snapshot = await fetcher().ConfigureAwait(false);
        _last[key] = snapshot;
        var delay = 0.0;
        if (snapshot.Error == ProviderErrors.RateLimited)
        {
            delay = 300;
        }
        else if (snapshot.Error == ProviderErrors.AuthRequired)
        {
            delay = 180;
        }

        _nextOk[key] = Environment.TickCount64 / 1000.0 + delay;
        return snapshot;
    }
}
